### PR TITLE
maps: Init `maps` module, add `filter`, `flat_map`, `maps` functions

### PR DESCRIPTION
**Pull request by ChAoSUnItY**
- Init `maps` module, an utility module works like `arrays` module but targets `map` data type.
- Add `filter`, `flat_map`, `maps`:
- `filter`:
- - Filters map entries by the given predicate function.
- `flat_map`:
- - Maps map entries into arrays and flattens into a one-dimensional array
- `map`:
- - Maps map entries into one-dimensional array
- Add tests:

```v
module maps

fn test_filter() {
	m1 := {
		0: 'ab'
		1: 'bc'
		2: 'cd'
		3: 'de'
		4: 'ef'
		5: 'fg'
	}
	assert filter(m1, fn (k int, v string) bool {
		return k % 2 == 0
	}) == {
		0: 'ab'
		2: 'cd'
		4: 'ef'
	}
	assert filter(m1, fn (k int, v string) bool {
		return v.contains('b') || v.contains('c')
	}) == {
		0: 'ab'
		1: 'bc'
		2: 'cd'
	}
}

fn test_maps() {
	m1 := {
		`a`: 'bc'
		`d`: 'ef'
		`g`: 'hi'
	}
	assert maps(m1, fn (k rune, v string) string {
		return '$k$v'
	}) == ['abc', 'def', 'ghi']
}

fn test_flat_map() {
	m1 := {
		1: [2, 3]
		4: [5, 6]
		7: [8, 9]
	}
	assert flat_map<int, []int, int>(m1, fn (k int, v []int) []int {
		mut a := [k]
		a << v
		return a
	}) == [1, 2, 3, 4, 5, 6, 7, 8, 9]
}

```

### CODE REVIEW

**Comment by Delta456** [comment]
Can we not have inbuilt methods for `flat_map`, `maps` and `filter` just like we have for array type?

**Comment by ChAoSUnItY** [comment]
my thought on this has two reasons:
1. It's quite a hard time to let them be inbuilt since it can makes more compiler magics happening
2. makes them inbuilt which indicates they're critical for the base of V lang, but I don't see any relatively critical reasons they should be in builtin

**Comment by JalonSolov** [comment]
The problem is that we would now have both `map` and `maps` types.  That could get confusing as to when one to use when.

Also, if it's going to be a separate type, it should probably go under `datatypes/maps` instead of directly in vlib.

**Comment by spytheman** [comment]
> The problem is that we would now have both `map` and `maps` types. That could get confusing as to when one to use when.

There is no separate `maps` type introduced by this PR - `maps` is a new module name, not a data type.

It just groups several generic functions working on maps, like `arrays` group several generic functions, that are useful for working on arrays.

**Comment by spytheman** [comment]
I disagree a little with the naming - if `maps.map` works like `a.map()`, it should produce a new map, that is the result of applying the function to each pair, similar to how maps.filter works, but just applying a function producing a new key, and a new value, instead of just returning a boolean.

**Comment by spytheman** [comment]
I agree that it is not good to implement these in `builtin` - Alex wants `builtin` to stay free of generics, and at the same time, adding more magic to the compiler does not look like a good idea either - there are many edge cases -> many future bugs for these special cases if we go that route. At least implementing it as generic functions reuses and tests common functionality that is used in many other places so the bugfixes will be shared :-| .

**Comment by spytheman** [comment]
> flat_map:
>    Maps map entries into arrays and flattens into a one-dimensional array
Given the description, perhaps `maps.to_arrays_and_flatten` ?

> map:
>    Maps map entries into one-dimensional array

Perhaps `maps.to_array` ?

**Comment by JalonSolov** [comment]
We already have .filter(), .map(), etc. on builtin types...

**Comment by spytheman** [comment]
> We already have .filter(), .map(), etc. on builtin types...

We do, and they are a constant source of bugs/edge cases, given how they are implemented in cgen - each uses unique compiler magic, generated code etc.

**Comment by spytheman** [comment]
Something like this, may be useful too:
```v
// to_map creates a new map, where each pair is produced by applying the function to a pair from the original map
pub fn to_map<K, V, X,Y>(m map[K]V, f fn (K, V) (X,Y)) map[X]Y {
    mut res := map[X]Y{}

    for k, v in m {
        x, y := f(k, v)
        res[x] = y
    }

    return res
}

fn test_map() {
    m := {'a':'b', 'c':'d'}
    res := to_map<string,string,string,int>(m, fn(k string, v string) (string, int) {
        return v, k[0]
    })
    dump(res)
}
```

It works and produces: `res: {'b': 97, 'd': 99}`, but I could not get it to infer the result pair X and Y types from the fn arg type for some reason.

**Comment by spytheman** [comment]
Also:
```v
pub fn invert<K, V>(m map[K]V) map[V]K {
    mut res := map[V]K{}
    for k, v in m {
        res[v] = k
    }
    return res
}

fn test_invert() {
    m := {10:'z', 20:'x'}
    res := invert(m)
    assert res == {'z': 10, 'x': 20}
    dump(res)
}
```

**Comment by spytheman** [comment]
Another one:
```v
pub fn from_array<V>(a []V) map[int]V {
    mut res := map[int]V{}
    for idx, v in a {
        res[idx] = v
    }
    return res
}

fn test_from_array() {
    assert from_array([20,30,40]) == {0: 20, 1: 30, 2: 40}
}
```

**Comment by ChAoSUnItY** [comment]
Ok, let me sum up the comment ideas here, so we will:
- Rename functions
- - `maps` -> `to_array`
- - `flat_map` -> `to_arrays_and_flatten`
- Adds new functions
- - `map`: Maps a map into a struct
- - `to_map`: Maps a map into map
- - `invert`: Maps map's entries, swap keys to entries and vice versa
- - `from_array`: Create a new map from array

Note: I am not sure if there's some compiler magic involved when parsing function named `map` which causes some unexpected error?

**Comment by spytheman** [comment]
> map: Maps a map into a struct

That is `to_map` from above, naming it just `map` does not work, because of the ambiguity with the `map` type.

**Comment by ChAoSUnItY** [comment]
Ok, I get it, I'll work on it.

**Comment by Delta456** [comment]
> I agree that it is not good to implement these in `builtin` - Alex wants `builtin` to stay free of generics, and at the same time, adding more magic to the compiler does not look like a good idea either - there are many edge cases -> many future bugs for these special cases if we go that route. At least implementing it as generic functions reuses and tests common functionality that is used in many other places so the bugfixes will be shared :-| .

I said it should be inbuilt because we already have an array map and filter methods so for consistency.

File: vlib/maps/maps.v [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+	for k, v in m {
+		x, y := f(k, v)
+		mp[x] = y
+	}
+
+	return mp
+}
+
+// to_map maps map entries by swapping key to value and vice versa
```
Comment (spytheman) [inline vlib/maps/maps.v:50]:
```suggestion
// invert returns a new map, created by swapping key to value and vice versa for each entry.
```

**Review by spytheman — APPROVED** [review]
